src/generic.rs

```diff
@@ -0,0 +1,110 @@
+use anyhow::{bail, Result};
+use bdk::{bitcoin::Network, database::MemoryDatabase, descriptor::get_checksum, Wallet};
+
+#[derive(Debug)]
+pub struct GenericXpub {
+    network: Network,
+    xpub: String,
+    derivation_path: String,
+    fingerprint: String,
+    first_address: String,
+    pub descriptor: String,
+}
+
+pub enum ScriptType {
+    Classic,
+    NativeSegwit,
+    WrappedSegwit,
+}
+
+fn script_type_from_path(path: &str) -> Result<ScriptType> {
+    let split_path: Vec<&str> = path.split("/").collect();
+
+    // TODO: learn a better Rust-ey way of handling this
+    let version = if split_path.get(1).is_some() {
+        split_path[1]
+    } else {
+        bail!("Something is wrong with this derivation path: {}", path)
+    };
+
+    let version = version.trim_end_matches(|c| c == '\'' || c == 'h');
+
+    match version {
+        "44" => Ok(ScriptType::Classic),
+        "49" => Ok(ScriptType::WrappedSegwit),
+        "84" => Ok(ScriptType::NativeSegwit),
+        _ => bail!("Didn't recognize the version number: {}", version),
+    }
+}
+
+fn build_descriptor(
+    xpub: &str,
+    derivation_path: &str,
+    fingerprint: &str,
+) -> Result<String> {
+    // m / purpose' / coin_type' / account' / change / index
+    // TODO: if network is regtest or testnet, make sure coin_type is 1
+    let origin_prefix = derivation_path.replace("m", &fingerprint.to_lowercase());
+    let hardened = origin_prefix.replace("'", "h");
```

Comment:
Consider combining rows as origin_prefix is never used, and you never need the version of the derivation path using commas again.

```suggestion
    let descriptor = derivation_path.replace("m", &fingerprint.to_lowercase()).replace("'", "h");
```